development/external-plugins/needs-tws/plugin.go

```diff
@@ -0,0 +1,305 @@
+		github.EventGUID, eventGUID,
+	)
+	defer l.Sync()
+
+	l.Debug("Got pull_request event")
+	var pre github.PullRequestEvent
+	if err := json.Unmarshal(payload, &pre); err != nil {
+		l.Errorw("Failed unmarshal json payload.", "error", err.Error())
+		return //return prematurely
+	}
+	if err := p.handlePullRequest(l, pre); err != nil {
+		l.Errorw("Failed to handle PR event", "error", err.Error())
+	}
+}
+
+func (p *PluginBackend) HandlePRReview(ep *externalplugin.Plugin, e externalplugin.Event) {
+	eventGUID := e.EventGUID
+	eventType := e.EventType
+	payload := e.Payload
+
+	l := externalplugin.NewLogger().With(
+		externalplugin.EventTypeField, eventType,
+		github.EventGUID, eventGUID,
+	)
+	defer l.Sync()
+
+	l.Debug("Got pull_request_review event")
+	var re github.ReviewEvent
+	if err := json.Unmarshal(payload, &re); err != nil {
+		l.Errorw("Failed unmarshal json payload.", "error", err.Error())
+		return //return prematurely
+	}
+	if err := p.handlePullRequestReview(l, re); err != nil {
+		l.Errorw("Failed to handle Pull Request Review event", "error", err.Error())
+	}
+}
+
+func (p *PluginBackend) handlePullRequest(l *zap.SugaredLogger, e github.PullRequestEvent) error {
+	// TODO (@Ressetkk): Configure this based on per-repository config
+	lb := DefaultNeedsTwsLabel
+
+	if e.Action == github.PullRequestActionClosed {
+		return nil
+	}
+	pr := e.PullRequest
+	if pr.Draft || pr.Merged {
+		return nil
+	}
+	org := e.Repo.Owner.Login
+	repo := e.Repo.Name
+	number := pr.Number
+
+	switch e.Action {
+	case github.PullRequestActionOpened, github.PullRequestActionReopened, github.PullRequestActionSynchronize:
+		var changed bool
+
+		changes, err := p.ghc.GetPullRequestChanges(org, repo, number)
+		if err != nil {
+			return err
+		}
+		for _, c := range changes {
+			if markdownRe.MatchString(strings.ToLower(c.Filename)) {
+				changed = true
+				break
+			}
+		}
+		labels, err := p.ghc.GetIssueLabels(org, repo, number)
+		if err != nil {
+			return err
+		}
+		hasLabel := github.HasLabel(lb, labels)
+
+		if !changed {
+			l.Debugf("Files not changed in %s/%s#%d@%s", org, repo, number, pr.Head.SHA)
+			if hasLabel {
+				l.Debug("remove stale label")
+				return p.ghc.RemoveLabel(org, repo, number, lb)
+			}
+			return nil
+		}
+		if hasLabel {
+			//we do not need to add another label
+			return nil
+		}
+		return p.ghc.AddLabel(org, repo, number, lb)
+	case github.PullRequestActionUnlabeled:
+		bc, err := p.ghc.BotUserChecker()
+		if err != nil {
+			return err
+		}
+		isBot := bc(e.Sender.Login)
+		if isBot {
+			// label removed by a bot. Skip.
+			return nil
+		}
+		deleted := e.Label.Name == lb
+		if !deleted {
+			// not a missing-docs label. Skip.
+			return nil
+		}
+		sender := e.Sender.Login
+		isCollaborator, err := p.ghc.IsCollaborator(org, repo, sender)
+		if err != nil {
+			return nil
+		}
+		if !isCollaborator {
+			// re-add label and notice only collaborators can bypass documentation review
```

Comment:
Why we allow collaborators to bypass documentation review?

Reply:
The plugin is meant to work when all collaborators and team members have their write access revoked, so the Prow is the one managing labels and Tide is merging PRs based on the state of labels. Note that this check will only work when the PR is **unlabeled*, that means the person is removing the label. This restricts to the people with the write access only.

If the people will need write access to the repo then for blocking purposes this plugin should not be used and the GitHub's `CODEOWNERS` should be used instead.